### ISSUE TITLE
[Feature Request] Easy failed update migration

### ISSUE THREAD

**Issue by theronnie602** [tachiyomiorg/tachiyomi#4616]
This is a QOL feature I'd like to see

I would like an option in updates in side the app to pull up the latest failed updates including whiy, from the last update only, instead of a text file I have to copy and paste the names from to migrate the sources in cases like "source URL has changed" or "no chapters found"

**Comment by curche** [comment]
'source URL has changed' and 'no chapters found' are issues related to particular sources
Refer the [troubleshooting guide](https://tachiyomi.org/help/guides/troubleshooting-problems/) and the [faq](https://tachiyomi.org/help/faq/#former-mangabox-extensions)

In such cases, you should be able to go to Browse > Migrate and view all manga under a single source and migrate them as you may require

**Comment by theronnie602** [comment]
There are some sources which will have 10-20 source URL changes at random times and when you have 300+ just from one source, it makes it a struggle and waste of time, plus  swapping back and forth from the word doc and app repeatedly can cause either to crash on some devices causing you to have to reupdate your entire library again,which for me already take about 30-40 mins, hence why I said it would be a QOL feature

**Comment by CrepeTF** [comment]
Source URL changes are very much mostly a mangabox source issue which we can do nothing about.

Regardless of having 300+ or 400+ or 1000+ manga from them, as long as you keep using it, you will have to migrate them anyway (whether it be back to the same source or another).

And this issue will persist as long as you persist to use them

You don't have to do them all at once.

**Comment by gp201** [comment]
> This is a QOL feature I'd like to see
> 
> I would like an option in updates in side the app to pull up the latest failed updates including whiy, from the last update only, instead of a text file I have to copy and paste the names from to migrate the sources in cases like "source URL has changed" or "no chapters found" 

I like this idea. Since some of my extensions keep changing the url. Everytime i have to search and migrate the manga to a new extension. 

Having a list within the app of all the failed updated and clicking on a chapter shows me all the possible migrations would be really helpful.

**Comment by uap4** [comment]
I vote that this is a real problem that needs to be solved. I already suggested for TachiyomiSY, but they also offered to manually migrate there. And in my case, with the URI changing even several times in the readmanga.live and mintmanga.live sources, it is practically impossible to track these silent changes manually. They are silent because the URI change, but when the old ones are requested, the response is redirected to the new URI. And the redirection happens without errors, so at first glance nothing breaks, although the old URI remains. But, if you do a search by source, because the URI are different, then what was found there, and what is also considered different in the library. And it makes no sense to think about the periodic migration of all thousands in the library. In my case, since a new URI is specified in the server's response, then instant migration as part of the metadata update is not so impossible. If at least it was possible to turn off redirection in okhttp3, then I would just sort the last update, find the ones that were not updated, select and run the migration. I could do that once a week.

So far I have tried to insert an http proxy into my chain of proxy, vpn and all sorts of things on the phone to drop redirects, but it turned out only once. And everything froze and rebooted, and then it didn't work anymore. Other options, like xposed or an extension fork, I think are even harder for me.

Sorry for the machine translation.

**Comment by CrepeTF** [comment]
> it is practically impossible to track these silent changes manually

well, not quite, there does exist a feature for tachi to send a you a notification of any library update errors and why

**Comment by uap4** [comment]
> notification of any library update errors

I repeat. At the moment, my case is not considered an error. Following the redirects, the program fully obtain all data. But this is if you look from the point of view of the library. And if you look at the source in the catalog or search, it can be seen, many manga in the library becomes orphaned. On READMANGA.Live and Mintmanga.Live, something changed probably, and the changes of the URI became madly a lot. Presumably that the URI will be changed from each active manga, and it is likely that even more than once.

But if the redirection leads to another site, then yes, it turns out a noticeable error, but if compared, then this is a small trouble.

**Comment by arkon** [comment]
Redirects can be disabled at an extension level. A few sources constantly changing their URLs doesn't justify app-wide disabling of redirections.

**Comment by uap4** [comment]
> Redirects can be disabled at an extension level

Probably, but this is in any case not user level. If there was a switch in the extension settings... I'll go there to write about it.

>A few sources constantly changing their URLs doesn't justify app-wide disabling of redirections.

And they were orphaned, too? In general, this is a workaround. And should strive to update the url like other metadata, and not break completely half-broken.

**Comment by T3sT3ro** [comment]
@valonsogit Implemented #8623, but closed it, so I think it is the appropriate place to once again raise that discussion. 

I really miss this feature. I've been missing it for more than 2 years, and what was implemented in the PR was the closest to the solution I've seen.

I think, it should be added either way even in the form that it was implemented in the PR. Having this feature in this state is 1000x better than not having it at all. For now I very frequently encounter the following problems and find myself struggling with how it's currently supposed to be solved:

- titles suddenly stop updating. On a regular basis, I have 20+ titles that just don't update anymore. It's harder too keep track of the more series I read concurrently.
- It is a lengthy and tedious process of fixing those errors:
  1. title X stops updating
  2. open the notification and error log in a separate app (HTML viewer, browser etc.), because tachiyomi doesn't have a dedicated window/modal for this log. It's also missing the "show latest errors" somewhere in the settings
  3. manually select a title, copy it, change app back to tachiyomi and paste in the library to look for a title and migrate it to another extension
  4. repeat for 20+ titles
- Titles require fixing at least several times a month

I understand, that maybe it's not the best way to implement it UX-wise, but at least it's better than the current way of showing a fail log in external app.

I would be in a strong favor of adding it to the app as-is for now, maybe with a small refactor, but also keep in mind that it's supposed to be refactored in the nearest future.

Also in terms of the comments to the PR and the UX, here are several of my thoughts:

- Titles failing seems to be an exceptional case in the app flow, so It should be treated as something requiring special attention from the user. That's why I would advise against it being just a filter in the library view. What it could be instead, is an interactive, red toast/notification (not sure how it's called) in the library screen, that would open a list of failed titles. 
- Displaying the error icon over the tittle is good, because it focuses user's attention on failing titles. It also encourages immediate action.
- Displaying error cause on the title, on the other hand, introduces too much clutter. If there was a list-view of failed titles, that could be a better place to display that information. Alternatively (or in addition to that) it could also be added to manga's page (where chapters are) and be visible under the title. The title could change the color to red in those cases.
- the failed state shouldn't be a filter. I think that filters should be used for the successful app flow, while title failing to update is an exceptional state. Otherwise it would be like going to a car saloon, and the tradesman telling you "Here are our red cars, here are diesels, and here are our stolen and wrecked cars". Additionally adding the "failed" category to the filters would hide that important information in the sub-menu 

For now I give you two concepts on how it could look like:

version A, toast:
![image](https://user-images.githubusercontent.com/5300963/233778469-d0b1561e-9e9a-4b68-a30d-56de50521c0d.png)

version B, sub-category in library view, errors on top with distinct error icons:
![image](https://user-images.githubusercontent.com/5300963/233778528-cad6e273-b7a7-478c-bf58-b29426fde6aa.png)

**Comment by T3sT3ro** [comment]
> The warning icon clutter is negligible since if you don't fix the error you aren't able to read the manga either way.

This is a very important point to consider, for anyone saying that the view is cluttered. I would honestly prefer cluttered view that helps me resolve the issue immediately, rather than obscured error hidden behind many layers. Functionality first, refining experience later.

I also completely agree that displaying error on manga page follows the fixing flow as @valonsogit described above.

Also I seem to have included wrong screenshot for the version A, so I'm fixing it now, check out the the updated previous message.

In terms of version B, I think that having the failed group at the bottom with a button to scroll would be a) to cluttered and b) hard to discover with a huge library. Instead, I propose the failed group be foldable and folded and collapsed default, so it doesn't take much space. Displaying the count next to it also seems like a good idea. See the updated image in previous message to see what I mean.

**Comment by arkon** [comment]
The suggestion of just adding it neglects the fact that it needs to be maintained, which makes adding anything to the library much harder. The library as-is is already complex with its various display modes and optional things that you can display on each item. You're going to be hard pressed to convince me that adding more things to the library, especially anything that entails actually breaking out of a regular grid/list, is a good idea.

I also agree that filtering does not really make sense, but at the same time scrolling through a large amount of items in a library to try to find erroneous items is also tedious, hence the desire to have a consolidated place to view them, which you can then go to the details/chapters view like you would from the library.

I'm envisioning literally just an interactive version of the existing plaintext content where you can navigate directly to the entries. I imagine that's what your Version A redirects to, which is essentially the same as what I had suggested from Updates, which logically seems related, in #8623.

**Comment by T3sT3ro** [comment]
Yes, the Version A would redirect to some other screen like an interactive version of the error list. Some things to think about though:
- sometimes a complete source goes obsolete, so it would be nice to have some grouping per-source like it is currently in the error log
- It would be good to display how many titles failed of the total amount of titles for a given source. 2/20 failed titles indicates that there are some problems. If 19/20 titles suddenly fail for a source, it is worth considering migrating everything from it.
- right now, error log is sorted primarily by error type and secondarily by source.

  
  This may make several sources occur in many different places which, I assume, doesn't really help extension developers nor users. It would be better to sort primarily by source, secondarily by name and only show error type as a secondary information under the title.

For the version B, the library view would gain additional level of nesting which is another level of complexity. The grouping into failed and regular, on the other hand, could be simpler to add then creating a whole nother screen and designing it from scratch. The grouping, in my opinion, works for any view: grid, condensed, cover only and list view. A question however remains, what happens when there are no errors: hide the subgroup entirely? It also pushes the content of main focus — the successfully updated library — slightly lower, which even now is pushed by the titlebar and categories by 1/5 of the screen height. Having it in the library, however, would allow reusing the same buttons for searching, filtering and refreshing both subgroups at the same time.

**Comment by arkon** [comment]
> The grouping into failed and regular, on the other hand, could be simpler to add then creating a whole nother screen and designing it from scratch. 

No it wouldn't. A new basic screen with a list is simpler than shoehorning a collapsible section into the library grid/list.

**Comment by Programmer-0-0** [comment]
I'll start working on this feature so if anyone has already started working on it, please inform me so we don't end up doing similar work :)

**Comment by quangkieu** [comment]
How about just a markdown viewer with redirect for #ID to comic ID?

**Comment by arkon** [comment]
A regular list would be more flexible in its functionality.

It could probably still follow the same hierarchy as we have in the text file using some paddings, with the actual entries looking like the library list view.

**Comment by Programmer-0-0** [comment]
Actually, I already started making a separate screen to display the failed updates manga. I'm almost done actually so I might create the PR in the next couple of days for you to review.

**Comment by Programmer-0-0** [comment]
@arkon I think I need your advice on this. Here is what I did so far:


<details>
<summary>Details</summary>

| The warning icon is disabled and not visible by default but when an update job is run and an error is detected it will become enabled| For now I only did it as a list (no grid view) and the migrate button has the same function as going to the migrate source screen|
| -------------   | ------------- |
![WhatsApp Image 2023-08-02 at 22 24 46](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/9c1d214e-a4a4-40fa-b438-405eb2ec7c9d)|                    ![WhatsApp Image 2023-08-02 at 22 24 46](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/df7b604c-26d0-47db-a455-d08194c2ed7a)
| I took @T3sT3ro advice and made the Pills to display the error count along side with the total count of mangas in that source| In the bottom bar there is a delete button as well as a dismiss button|
![WhatsApp Image 2023-08-02 at 22 24 48](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/d6ecd086-e9ad-4433-9abd-1c6c98f65e76) |        ![WhatsApp Image 2023-08-02 at 22 24 50](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/61f03a35-c6eb-4630-b0f9-00d1e7dc2a10)    
| There is also an option to group by category but I couldn't find a way to efficiently get the total manga count in that category   |I edited the EmptyScreen function to display happy faces when you dismiss or ignore all the errors|
![WhatsApp Image 2023-08-02 at 22 34 39](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/9f8588a4-037b-4fba-aebe-c2aaedb0d881) | ![WhatsApp Image 2023-08-02 at 22 34 39](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/08a44eff-67a7-4634-b7c1-50c718448b8d)

</details>
               
The main problem right now is the error message. It's a bit long so I wanted some advice on how to tackle it. I thought of creating a simplified user friendly error message by catching different types of Network exceptions and if the user wants to see the error in detail he could open the txt file instead. Although one work around is to rotate the screen:

<details>
<summary>Image</summary>

![WhatsApp Image 2023-08-02 at 22 24 48](https://github.com/tachiyomiorg/tachiyomi/assets/67505416/ce63c82f-698c-435d-b3c8-7a544f5b26d1)

</details>

**Comment by THElegend5** [comment]
> EDIT: other than that, I'm basically done. I just need to refactor the code to prepare it for the PR so if anyone has any suggestions please let me know.

Maybe this along with #9737 and  
#9623,  #7502 (when they are implemented) be clubbed under a library management tab in more? As you wouldn't really look at these that freqently and would allow to view any library anomaly in one place.

**Comment by Programmer-0-0** [comment]
@THElegend5 I can agree that you wouldn't look that frequently at #9737 and #9623 but I don't think that's the case for the failed updates list, the people who have 200+ manga in their library are bound to get at least 1 failed update which they need to fix several times a month. Personally, I rarely go the more tab (maybe that's the same for other people) so putting this feature there will be kind of tedious compared to putting it in the updates screen where you mainly see the updates. Also, the mechanism of the warning icon is that it won't appear unless you have any failed updates (it gets disabled on each library update job and if an error is found it gets re-enabled) I did that to reduce the bloating since @arkon mentioned it in one of his previous posts.

**Comment by arkon** [comment]
> The main problem right now is the error message. It's a bit long so I wanted some advice on how to tackle it.

I may be biased by the existing implementation, but grouping by error message, then source, then the individual entries helps with that, assuming you allow multiple lines for the error message.

**Comment by T3sT3ro** [comment]
From user standpoint I think that grouping priority should not be error > source > title by default, but either by source > title or title > source. There could be an option to switch to the error > source > title, but from experience this is not useful for a user, but rather for developer. Generally as a user I recognize these types of error: 

- "temporary manga problem, try to refresh"
- "permanent manga problem, url migrated or something, manga has to be migrated"
- "temporary source problem, update was unsuccessful"
- "permanent source problem, you have to migrate from that source"

As for displaying error message, I would add a side-to-side ping-pong or scroll animation for the text message. I would also add a notification directly on the manga page, possibly over the manga synopsis notifying about the last update error for that source. Maybe instead of a message in a fixed place I would use a dismissable snackbar with warning and error message instead.

**Comment by Programmer-0-0** [comment]
> I may be biased by the existing implementation, but grouping by error message, then source, then the individual entries helps with that, assuming you allow multiple lines for the error message.

Do you mean group by source then by error? because if you group by error first it will be weird as different manga entries under the same source may have different types of errors (I've encountered such cases many times before). Also, I already tried allowing multiple lines while debugging but it didn't look that attractive.

**Comment by stevenyomi** [comment]
The current text implementation is "error > source > titles" but I also think "source > error > title" might be better.

Most of the suggestions from @T3sT3ro are too bloat (especially adding options). This is mentioned by arkon in this thread many times. And setting error to marquee ("side-to-side ping-pong or scroll animation") is a very bad idea, as user can hardly read it or screenshot it, and may cause headache for support people.

**Comment by Programmer-0-0** [comment]
> As for displaying error message, I would add a side-to-side ping-pong or scroll animation for the text message. I would also add a notification directly on the manga page, possibly over the manga synopsis notifying about the last update error for that source. Maybe instead of a message in a fixed place I would use a dismissable snackbar with warning and error message instead.

I'm not quite sure what you mean by side to side text message, do you mean a button on the manga entry you can press then the error message will display in an animated style? Also the dismissible snackbar (assuming you mean to each individual manga screen) isn't a bad idea but then you would need to display some sort of icon on the manga entry in the library so you can know that that manga has an error (which increases the bloat) or did you mean display all the errors in one dismissible snackbar? Honestly, I was thinking of gathering the most common errors and converting them to user-friendly readable error messages and displaying them and then if an uncommon error appears the text "Press error details button for more info" and add a button to display the full error message in the bottomActionMenu

**Comment by stevenyomi** [comment]
[Marquee](https://www.geeksforgeeks.org/how-to-create-marquee-text-in-android/) (don't do this)

> Honestly, I was thinking of gathering the most common errors and converting them to user-friendly readable error messages and displaying them and then if an uncommon error appears the text "Press error details button for more info" and add a button to display the full error message in the bottomActionMenu

No, don't do that. Keep things as simple as possible.

**Comment by Programmer-0-0** [comment]
> The current text implementation is "error > source > titles" but I also think "source > error > title" might be better.

Alright so do I try to make it "source > error > title" and "category> error > title" and the ungroup option will display all the manga with the error message as secondary info like how it is currently?

The problem is still with the error message being too long to display in the header, we need to deal with it in a smart way.

>[Marquee](https://www.geeksforgeeks.org/how-to-create-marquee-text-in-android/) (don't do this)

lol noted.

**Comment by stevenyomi** [comment]
> That's exactly the intention of grouping by errors first. It's easier to determine an action based on an error rather than seeing the same thing again under multiple sources.

I wonder in what situation would the same error message show up in multiple sources. A counter-example is when you don't have internet connection - in this case each source would have different error messages containing their respective domains, as is shown in https://github.com/tachiyomiorg/tachiyomi/issues/4616#issuecomment-1662821422.

**Comment by T3sT3ro** [comment]
I honestly don't find grouping by error first useful for users, My options suggestion was to accommodate for arkon's prior bias, as he called it. I think that even for contributors and extension developers the source > error > title is more relevant, because they only work on one extension at a time. As a user I don't do action per error first but rather per source first.

@Programmer-0-0 Yeah, I meant Marque, and I don't think that's such a bad idea:

- The full error message is already included in the logs, if someone needs it. There should be no need for user to screenshot the error message. It's a bad practice anyway, because the message can't be later copy-pasted and must be handwritten when trying to debug anyways.
- Marques move slowly, reading the text shouldn't be a problem in normal scenario.
- The whole point of this discussion will be obsolete if the source > error > title grouping would be present. Still valid for the "ungrouped" view though.
- However as @stevenyomi noted Marques can play bad with "support people" (you meant from accessibility standpoint?), but what concrete examples do you have in mind?

I mean I'm also not sold on my own idea about Marque, but I'd like to know (for my own future reference as well) what are some concrete scenarios when it's bad.

@Programmer-0-0 As for the snackbar — I meant inside the manga title, a single snackbar with message like "Last manga update failed:" and an error message in the next line ("Unable to resolve host"... etc.) for that title only. 

Is the icon on the library page needed? I don't know, I don't think so, but it can indeed be useful at times. With your error list those icons could be redundant, no? The snackbar however would be useful for any navigation path that does not start from your error list, like coming from updates screen by clicking cover icon to go to title, browsing mangas on search screen etc. It would also help when trying to refresh a single manga only and it fails.

Lastly, unrelated to the previous, I think you should add an option to open the error log file to the "more" menu on the failed updates screen. Currently opening the log via notification is tedious, because after dismissing it you have to look through phone files in some file explorer to access it and you have to trigger the whole library update manually.

**Comment by stevenyomi** [comment]
> The full error message is already included in the logs, if someone needs it.

No, the new screen is to completely replace that log.

> Marques move slowly, reading the text shouldn't be a problem in normal scenario.

Even the marquee moves slowly, it still makes the screen looks messy. It's like many error messages flying around at the same time.

If it moves very, very slowly, it would be very inefficient for people who read very fast.

> Marques can play bad with "support people"

Screenshots with truncated error messages will show up everywhere in GitHub issues and Discord support channel. "Support people" refers to contributors who reply to those messages.

> Is the icon on the library page needed?

Already rejected in #8623.

**Comment by T3sT3ro** [comment]
I think it absolutely should **_NOT_** replace the log but be a primary way to view errors from inside the app.

**Comment by T3sT3ro** [comment]
For reporting on github for example. Will you re-write a potentially log error message or stack trace from a screenshot?

**Comment by stevenyomi** [comment]
The error log file only has the same single-line error message. Stacktraces are not logged in it. In fact it's not logged anywhere, unless the extension intentionally caught it and logged it down.

**Comment by AntsyLich** [comment]
> For reporting on github primarily. Will you re-write a potentially log error message or stack trace from a screenshot? How will you report to extension developer that manga can't update without the log?

Add tap to copy function to error message. Profit $$$

**Comment by Programmer-0-0** [comment]
> Add tap to copy function to error message. Profit $$$

probably the easiest thing to implement and we can all agree on until now lol

**Comment by arkon** [comment]
- Do source then error grouping
- UI replaces log file entirely
- Allow multiline error messages regardless of how bad it currently looks. We can iterate on that if needed.
